engine/class_modules/sc_demon_hunter.cpp

```diff
@@ -4723,6 +4723,49 @@ struct vengeful_retreat_t : public demon_hunter_spell_t
   }
 };
 
+// Soul Carver =========================================================
+struct soul_carver_t : public demon_hunter_attack_t
+{
+  struct soul_carver_oh_t : public demon_hunter_attack_t
+  {
+    soul_carver_oh_t( demon_hunter_t* p )
+      : demon_hunter_attack_t( "soul_carver_oh", p, p->talent.vengeance.soul_carver->effectN( 3 ).trigger() )
+    {
+      background = dual = true;
+      may_miss = may_parry = may_dodge = false;  // TOCHECK
+    }
+  };
+
+  soul_carver_oh_t* oh;
+
+  soul_carver_t( demon_hunter_t* p, util::string_view options_str )
+    : demon_hunter_attack_t( "soul_carver", p, p->talent.vengeance.soul_carver, options_str )
+  {
+    oh        = new soul_carver_oh_t( p );
```

Comment:
I would do this as
p->get_background_action<soul_carver_oh_t>( "soul_carver_oh" );

And change the signature above to:
  soul_carver_oh_t( util::string_view name, demon_hunter_t* p )
      : demon_hunter_attack_t( name, p ... )
      
Additionally, you could assign this as 
impact_action = p->get_background_action ... 

That way you don't need to do the manual execute below.